Content validation design doc (#9658)
department-of-veterans-affairs/va.gov-team: platform/engineering/design-docs/2020-05-26-content-validation.md

This design document is intended for front end and DevOps engineers on the Veteran-facing Services Platform (VSP) and Veteran-Facing Services (VFS) teams.

### Background

The `vets-website` repo currently contains [one script that builds both the content and the applications](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/index.js). Chris V. wrote a [design doc to disentangle the content build from the application build](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/engineering/design-docs/2020-04-09-separate-content-build.md), and that work is currently in progress.

Right now, content validation consists of broken link checking and accessibility checking.

##### Broken Link Checking

The `check-broken-link` plugin does the following:

1. Loop through all of the HTML files in the Metalsmith pipeline
2. Extract all of the broken `href`/`src` values using the `getBrokenLinks` helper
3. Formats the result into useful console output
4. Breaks the build on production, and logs the output on lower environments

##### Accessibility Checking

> 

Author: When the integration step of the build fails, there is a generic notification sent to the #devops-deploys channel via https://github.com/department-of-veterans-affairs/vets-website/blob/f29aa3a385d03479308f78c0c22c292ecae93f20/Jenkinsfile#L86, correct? If so, the violation must be reported manually because the error message isn't descriptive, correct?

For example: https://dsva.slack.com/archives/C37M86Y8G/p1591099447374800

> 

Reviewer: I believe that's all correct, yeah.

#### Pain points

##### Content Writers

Content writers want to quickly draft, publish, and deploy content. However, `vets-website` is only deployed to production once per day, and `vets-website` deploys to staging take 20+ minutes. To allow content writers to move faster, there is a partial deploy, which only includes static page changes (`vagov-content` and Drupal).

> Unfortunately, the partial deploys have no content validation. That means that content writers don't have a proactive reactive way to learn if their content includes broken links or accessibility errors.

Reviewer 2: Content writers do have validation from the node link report.

Where it still has the potential to fall down right now is that  if a path to a page in drupal is changed, drupal automatically creas a redirect from the old path to the new path, so when the links are checked, they still work.   The original plan was that the Front end would import redirects from drupal into Nginx when the content build takes place.  However that is not happening, so the front end does not have the redirects, so a moved page, results in broken links from pre-existing content links.

> Unfortunately, the partial deploys have no content validation. That means that content writers don't have a proactive reactive way to learn if their content includes broken links or accessibility errors.

Reviewer 3: So many of the broken links come from URLs that change, according to decisions by the Content & IA team, usually in conjunction with Public Websites. In that workflow, an engineer (usually @ncksllvn, i believe, but others seem to do it too) makes a code change for the redirect, and has to time the release of the redirect code to match the content release of the URL change in the content layer. It's a fragile and very time-consuming process. 

At one point there was [an epic on the operations team's backlog to establish Drupal as source of truth for redirects](https://github.com/department-of-veterans-affairs/va.gov-team/issues/503). but i just saw that the initial progress towards that was just removed in [PR 6999](https://github.com/department-of-veterans-affairs/devops/pull/6999)

On our end, we have [a new user story](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/1937) that may help a bit with this, much like the "outgoing" node link report mentioned by Steve. But these solutions rely heavily on humans to do things in the right order for things to not fall down.

> Unfortunately, the partial deploys have no content validation. That means that content writers don't have a proactive reactive way to learn if their content includes broken links or accessibility errors.

Reviewer 3: I should add, as we add hundreds of editors from outside the tight Github/Slack communication flows (eg Public Affairs Officers at 2000+ VA facilities), we won't be able to rely on this type of coordination for redirects.

> Unfortunately, the partial deploys have no content validation. That means that content writers don't have a proactive reactive way to learn if their content includes broken links or accessibility errors.

Author: @swirtSJW @kevwalsh, thank y'all for the feedback so far. I'll let you know once I'm done updating the design doc to reflect this improved understanding.

##### Users

> Content-only deploys create an opportunity for end users to encounter broken links and/or accessibility errors from CMS content that wasn't validated before being deployed.

Reviewer: Expand on this a little. This happens because content can bypass any content validation checks if it's deployed via the content-only deployment.

> ##### FE Tools team

Reviewer: The scope is much broader here. This affects all teams across all contracts who are trying to ship code on the VA.gov front end, not just the FE Tools team.

## Specifics

### Detailed Design

COMING SOON

### Code Location

The validation of the application content will remain in `vets-website`. The validation of the static content will live in the [new `content-build` repo](https://github.com/department-of-veterans-affairs/content-build).

### Testing Plan

COMING SOON

### Logging

COMING SOON

### Debugging

COMING SOON

### Caveats

- Our [function to determine if a link is broken](https://github.com/department-of-veterans-affairs/vets-website/blob/e8b68850fc83fdae4386fe4ab392770dde38faee/src/site/stages/build/plugins/check-broken-links/helpers/isBrokenLink.js#L16) doesn't check if external links are broken.
- Our [accessibility check](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/assets/js/execute-axe-check.js#L8) only renders a banner on the preview server.

### Security Concerns

There are no new security concerns with a separated build process.

### Privacy Concerns

There are no new privacy concerns with a separated build process.

### Open Questions and Risks

COMING SOON

### Work Estimates

COMING SOON

### Alternatives

#### Highest quality, but lowest speed

> If we could proactively validate CMS content before it's published, we could eliminate a source of downstream broken links and/or accessibility errors, which would improve the experience for end users and/or reduce the number of frontend deploys that require manual intervention. The downside of this approach would be figuring out how to add the content validation to the CMS, and/or adding friction to the content writers workflow.

Reviewer 2: @cvalarida We added link checking to the CMS several months back On node save every link both internal and external are tested and a report generated for that node.   
![image](https://user-images.githubusercontent.com/5752113/83689439-0a64c680-a5bd-11ea-9e38-e5c855f6f78e.png)
A bright red box appears if there are any that are broken.
![image](https://user-images.githubusercontent.com/5752113/83689693-79dab600-a5bd-11ea-81bb-58755dd95bbe.png)

508 compliance is a little tougher because the front end templates do not match the cms templates,

> If we could proactively validate CMS content before it's published, we could eliminate a source of downstream broken links and/or accessibility errors, which would improve the experience for end users and/or reduce the number of frontend deploys that require manual intervention. The downside of this approach would be figuring out how to add the content validation to the CMS, and/or adding friction to the content writers workflow.

Reviewer 2: This checker even reaches out to the public website in case it finds pages that are not known to drupal.  

All contributed back to benefit Drupal's everywhere :)  
https://www.drupal.org/project/node_link_report

> If we could proactively validate CMS content before it's published, we could eliminate a source of downstream broken links and/or accessibility errors, which would improve the experience for end users and/or reduce the number of frontend deploys that require manual intervention. The downside of this approach would be figuring out how to add the content validation to the CMS, and/or adding friction to the content writers workflow.

Reviewer 2: It is on my roadmap to add link accessibilty checking to the link checker which would warn on an editor creating a link with no text in it.   This was a recent issue that blocked a content release a couple of weeks back.
https://www.drupal.org/project/node_link_report/issues/3145477

> If we could proactively validate CMS content before it's published, we could eliminate a source of downstream broken links and/or accessibility errors, which would improve the experience for end users and/or reduce the number of frontend deploys that require manual intervention. The downside of this approach would be figuring out how to add the content validation to the CMS, and/or adding friction to the content writers workflow.

Reviewer 4: @swirtSJW how come when we see reports of broken link they seem to originate from failing builds? 

If we're confident with Drupal's link validation, then I think we should just delete link validation from vets-website.

> If we could proactively validate CMS content before it's published, we could eliminate a source of downstream broken links and/or accessibility errors, which would improve the experience for end users and/or reduce the number of frontend deploys that require manual intervention. The downside of this approach would be figuring out how to add the content validation to the CMS, and/or adding friction to the content writers workflow.

Reviewer 2: @ncksllvn I think most now are either related to something being unpublished or moved to a new path.  The node link report warns if linking to something that is unpublished, but it does not / can not prevent it.

> If we could proactively validate CMS content before it's published, we could eliminate a source of downstream broken links and/or accessibility errors, which would improve the experience for end users and/or reduce the number of frontend deploys that require manual intervention. The downside of this approach would be figuring out how to add the content validation to the CMS, and/or adding friction to the content writers workflow.

Reviewer 2: Also, our link validation happens on a save in drupal.  It is not blocking, only reported.   It is up to editors to note and fix.

### Background

The `vets-website` repo currently contains [one script that builds both the content and the applications](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/index.js). Chris V. wrote a [design doc to disentangle the content build from the application build](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/engineering/design-docs/2020-04-09-separate-content-build.md), and has since broken the build into two two separate stages.

1. [Run Webpack](https://github.com/department-of-veterans-affairs/vets-website/blob/43ea0bdccd5e53886e3e38ea27d3f8e8e7bd9038/script/build.sh#L55)
2. [Run the content build](https://github.com/department-of-veterans-affairs/vets-website/blob/43ea0bdccd5e53886e3e38ea27d3f8e8e7bd9038/script/build.sh#L61)

Previously, Webpack was entangled in the content build, but no longer. Brooks is currently working on moving the content build into a [`content-build`](https://github.com/department-of-veterans-affairs/content-build/) repo to separate it from `vets-website`.

##### Content Sources

Static content currently comes from two sources:

1. The Drupal CMS

> 2. The [`vagov-content` repo](https://github.com/department-of-veterans-affairs/vagov-content/blob/master/pages/404.md)

Reviewer 4: Also - we still have static content in vets-website - the entire `src/site/assets` directory iirc

This design doc will focus on validating content from the Drupal CMS.

##### Amazon S3

The frontend of VA.gov is served as static files via Amazon S3. Since `vets-website` currently handles both application and content building, a failure in one blocks both from being deployed to S3 during a full deploy.

###### Broken Link Checking

> 

Reviewer 5: summarize what this validates

The [`check-broken-link` middleware](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/plugins/check-broken-links/index.js) is a Metalsmith plugin that happens near the end of the Metalsmith static content build pipeline. It does the following:

1. Loop through all of the HTML files in the Metalsmith pipeline
2. Extract all of the broken `href`/`src` values using the `getBrokenLinks` helper
3. Formats the result into useful console output

> 4. Blocks the deployment on production or logs the error output on lower environments

Reviewer 4: Some historical info might be useful - originally (on vets.gov), all content was in vets-website in a `content/` directory and you would open a PR to edit content. CI would then run against that PR. Link-checking would run as part of the build the same way it does today. If a broken link was found, CI would fail, and you wouldn't be able to merge. This meant that broken links never made their way into the master branch.

Accessibility checks had the same deal - errors never made their way into master.

Also - I think Drupal contains some link-checking internally nowadays but it isn't as comprehensive or something.

###### Accessibility Checking

> 

Reviewer 5: summarize what this validates

The accessibility checks only happen in CI ([the Integration step in Jenkins](https://github.com/department-of-veterans-affairs/vets-website/blob/master/Jenkinsfile#L81), which only happens during the full `vets-website` build.

Before accessibility checks can be run, the sitemap is generated via the [`create-sitemap` Metalsmith plugin](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/plugins/create-sitemaps.js). Nightwatch iterates over all the paths found in the sitemap, and runs the [aXe accessibility checker](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/testing/e2e/nightwatch-commands/axeCheck.js) to look for violations.

When violations are found, the build breaks and violations must be reported manually.

> The accessibility checks are skipped during a content-only deploy.

Reviewer 4: All tests (except link checking) are skipped during content-only deploys by design - for speed of publishing. The link-checking runs as part of the build itself rather than as a separate step whereas unit tests run before before e2e and accessibility tests afterwards.

> The accessibility checks are skipped during a content-only deploy.

Reviewer: :point_up: That context is probably good to add to the doc, @billfienberg .

Additionally, we have the preview server which shows content editors any accessibility violations, but they must deliberately go to the server from a link in Drupal to see this.

#### Pain points

##### Content Writers

> Content writers want to quickly draft, publish, and deploy content. However, `vets-website` is only deployed to production once per day, and `vets-website` deploys to staging take 20+ minutes. To allow content writers to move faster, there is a partial deploy, which only includes static page changes (`vagov-content` and Drupal).

Reviewer 4: vets-website deploys take 20 mins to go to staging bc CI runs again on master - the deploy to staging itself takes like a minute.

But also, staging.va.gov has pretty much zero value to content writers, so it's probably not even worth mentioning. Writers are editing in Drupal prod and looking at their changes via the preview server. So they are basically always in prod. Drupal staging is wiped by a downstream migration from Drupal prod every now and then, don't know when that job runs. But it basically means that Drupal staging reflects old Drupal prod content and any changes you make are temporary, so basically, there is no value to anyone to make content changes to Drupal staging.

(related but not completely on topic - this whole environment config between staging.va.gov and drupal staging has been a serious pain point for app teams, but i think folks roll their eyes now whenever i bring it up 🙄 )

Reviewer 2: CMS staging and dev are rebuilt every time a commit is merged into master.   When it is rebuilt, it pulls a new db copy from prod.

> Unfortunately, the partial deploys have no content validation. That means that content writers don't have a proactive way to learn if their content includes broken links or accessibility errors.

Reviewer 4: Does ` partial deploy` == content deploy?

Someone should definitely confirm broken links don't block content deploys because I thought they did.

Reviewer: Oh hey, look at that. So they do.
![image](https://user-images.githubusercontent.com/12970166/83803417-5c5f2800-a661-11ea-98ec-117a8f44769a.png)
From `checkForBrokenLinks` in `jenkins/common.groovy`.

I even made that change, but I totally forgot. My mistake! I've led you astray, @billfienberg.

Author: When I say `partial deploy`, I am referring to the content deploy. I got the term `partial deploy` from the [Separate Content Build design doc](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/engineering/design-docs/2020-04-09-separate-content-build.md#deployment), which links to the [Deploy page of the VA.gov workflow docs](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/getting-started/workflow/deploy/#partial-deploy--static-page-changes-only).

Author: Oh yea. There it is in `common.groovy`.

```
    // Only break the build if broken links are found in master
    if (IS_PROD_BRANCH || contentOnlyBuild) {
      throw new Exception('Broken links found')
    }
  } else {
    echo "Did not find broken links."
  }
```

https://github.com/department-of-veterans-affairs/vets-website/blob/master/jenkins/common.groovy#L177

##### Users

> Content-only deploys create an opportunity for end users to encounter broken links and/or accessibility errors. This happens because content can bypass any content validation checks if it's deployed via the content-only deployment.

Reviewer 4: Is this true for broken links? See my earlier comment - the broken links run as part of the build (website compilation), not as an isolated step. Do we skip link-checking during content deploys? We would have had to do that via a flag passed into the build command or something. Seems like a bad idea.

Reviewer: Right, yeah. Looks like broken links are accounted for after all. I gave Bill the wrong information there. :raised_hand_with_fingers_splayed: : My bad.

The accessibility checks aren't run, but broken links are checked.

##### FE Tools team

Invalid content causes warning logs during staging deploys, and breaks the build during production deploys. These broken builds then need manual intervention, and require coordinating with another team.

##### Anybody that is trying to ship code for the VA.gov front end

If invalid content is breaking the production build, and therefore blocking the production deploy, then no team on any contract is able to ship their frontend code to VA.gov.

### High Level Design

Once the `content-build` repo is successfully extracted from `vets-website`, the Metalsmith static content build pipeline will be deleted from `vets-website`. Webpack will handle the application build in `vets-website`, and Metalsmith will handle the static content build in the `content-build` repo. This will create two separate pipelines for built code to be deployed to S3, which will prevent failures in one from blocking the other.

> The validation of the static content coming from the CMS will occur during the Metalsmith build.

Reviewer 4: > The validation of the static content coming from the CMS will occur during the Metalsmith build.

I think what we're going for here is a separation of concerns or delineation of responsibilities, but so we're clear, engineers will always need to support the content writers. It's cool that we're going for content validation not holding up app deploys, but also, we can't just put the content layer aside and act like it isn't our problem anymore.

## Specifics

### Detailed Design

> The content validation will happen in a daily scheduled job.

Reviewer 4: To run in a daily scheduled job, I assume that it's report-only at that point? Because as an isolated job, it won't be blocking any deployments.

> Currently, the static CMS content is only validated during the full deploy, which only happens once per day. To maintain some form of parity with that that process, the content validation will happen in a daily scheduled job.

Reviewer 5: This section should have a lot more information on the implementation. What is going to run? What is going to be tested? Is there documentation on the existing process that can be referenced? What are the inputs and outputs? 

If you've already documented this in the background, then reference it here. something like, "the existing link checking process will be used. it will be triggered by a daily job blah blah."

### Testing Plan

1. Configure a scheduled job to run a script once per workday
2. Write an initial script that logs when the script is run
3. Once the scheduled job is running at the correct time, replace the initial script with the actual script that runs the content validation

### Logging

Currently, invalid content is logged during staging deploys. And failed builds trigger generic [Slack notifications](https://dsva.slack.com/archives/C37M86Y8G/p1591099447374800).

### Debugging

##### Broken links

1. Confirm that the content build is passing
2. Modify an HTML file, like [`header.html`](https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/includes/header.html) to have a broken link
3. Run the build again to confirm that it is failing

##### Accessibility errors

1. Confirm that the content build is passing
2. Modify an HTML file, like [`header.html`](https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/includes/header.html) to have an accessibility error
3. Run the build again to confirm that it is failing

### Caveats

> - Our [function to determine if a link is broken](https://github.com/department-of-veterans-affairs/vets-website/blob/e8b68850fc83fdae4386fe4ab392770dde38faee/src/site/stages/build/plugins/check-broken-links/helpers/isBrokenLink.js#L16) doesn't check if external links are broken.

Reviewer 4: I don't think this this worth mentioning really because the proposed design here doesn't change this behavior. We've never validated broken links.

- Our [accessibility check](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/assets/js/execute-axe-check.js#L8) only renders a banner on the preview server.

### Security Concerns

There are no new security concerns with a separated build process.

### Privacy Concerns

There are no new privacy concerns with a separated build process.

### Open Questions and Risks

#### Risks

- Neither the current content validation nor the proposed content validation will **prevent** invalid content from making it to production. But this proposal would prevent unrelated teams from being blocked by invalid content.

#### Questions

### Work Estimates

The following estimates vary greatly depending on who's doing the work.

> 1. Finish extracting the `content-build` repo out of the `vets-website` repo - ?

Reviewer 4: I'm assuming the content-build repo will need to be sibling to vets-website at all times, correct? CMS content is everywhere, from the mega menu to user flows. Does this mean vets-website, content-build, and vagov-content will all need to be cloned side by side?

Reviewer 6: Yes, that's correct.

##### Content Sources

Static content currently comes from two sources:

1. The Drupal CMS
2. The [`vagov-content` repo](https://github.com/department-of-veterans-affairs/vagov-content/)

This design doc will focus on validating content from the Drupal CMS.

##### Amazon S3

The frontend of VA.gov is served as static files via Amazon S3. Since `vets-website` currently handles both application and content building, a failure in one blocks both from being deployed to S3 during a full deploy.

###### Broken Link Checking

The [`check-broken-link` middleware](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/plugins/check-broken-links/index.js) is a Metalsmith plugin that happens near the end of the Metalsmith static content build pipeline. It does the following:

1. Loop through all of the HTML files in the Metalsmith pipeline
2. Extract all of the broken `href`/`src` values using the `getBrokenLinks` helper
3. Formats the result into useful console output

> 4. Blocks the deployment on production or logs the error output on lower environments

Author: Slack notification sent to the #cms-team channel informing them of broken links found in the build. 

https://dsva.slack.com/archives/CT4GZBM8F/p1591303292290600

![image](https://user-images.githubusercontent.com/6130520/83812868-def7df80-a681-11ea-904a-b9bd62ea8b67.png)

If invalid content is breaking the production build, and therefore blocking the production deploy, then no team on any contract is able to ship their frontend code to VA.gov.

### High Level Design

Once the `content-build` repo is successfully extracted from `vets-website`, the Metalsmith static content build pipeline will be deleted from `vets-website`. Webpack will handle the application build in `vets-website`, and Metalsmith will handle the static content build in the `content-build` repo. This will create two separate pipelines for built code to be deployed to S3, which will prevent failures in one from blocking the other.

The validation of the static content coming from the CMS will occur during the Metalsmith build.

## Specifics

### Detailed Design

The content validation will happen in a daily scheduled job.

### Code Location

The content validation script and scheduled job config will live in the [new `content-build` repo](https://github.com/department-of-veterans-affairs/content-build).

### Testing Plan

1. Configure a scheduled job to run a script once per workday
2. Write an initial script that logs when the script is run

> 3. Once the scheduled job is running at the correct time, replace the initial script with the actual script that runs the content validation

Reviewer 6: This feels a bit confusing to me.  Can you try rewording it or explaining it differently?

### Logging

> Currently, invalid content is logged during staging deploys. And failed builds trigger generic [Slack notifications](https://dsva.slack.com/archives/C37M86Y8G/p1591099447374800).

Reviewer 6: Do you plan on changing this?

Author: The new content validation script will continue including logs in the build log. On top of that, it will send an automated report of invalid content to the #cms-team Slack channel.

### Debugging

##### Broken links

1. Confirm that the content build is passing
2. Modify an HTML file, like [`header.html`](https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/includes/header.html) to have a broken link
3. Run the build again to confirm that it is failing

##### Accessibility errors

1. Confirm that the content build is passing
2. Modify an HTML file, like [`header.html`](https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/includes/header.html) to have an accessibility error
3. Run the build again to confirm that it is failing

### Caveats

> - Our [accessibility check](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/assets/js/execute-axe-check.js#L8) only renders a banner on the preview server.

Reviewer 6: This seems like it could easily get confused with the accessibility tests that Jenkins runs.  Can you differentiate them a bit more?

Author: I actually removed this caveat because I added it before I realized that there are actually two different types of a11y checks. 

1. During the Metalsmith build
2. During the Integration stage of Jenkins

Reviewer 4: FYSA -

> During the Metalsmith build

This is totally non-blocking and runs only on the preview server. It _can_ be run during a regular build or watch task, but that's not what it was designed for.

> During the Integration stage of Jenkins

This _does_ block regular full deployments. The gist is that -

1. Everything is good and CI is passing
1. A new page is published in Drupal
1. CI runs on vets-website master or on a PR branch
1. The new page from Drupal is served into vets-website CI
1. The new page has an accessibility issue, and now all branches in vets-website have to fall back to the cached Drupal data on S3 in order to build successfully.
        - This doesn't always work. Sometimes the cache is considered invalid if the GraphQL queries have changed since the last cache was uploaded, so if there's an accessibility issue at the same time, CI is blocked for all branches.
1. Full deployments can't go out. 
    - Content-only deploys skip accessibility tests, so content can (ironically) still go out.
    - If vets-website master passed okay during the last build, we can do an app-code deployment by skipping the pre-release job in Jenkins. (Just unchecking the box for the deploy job)

So...we could technically achieve a full deployment by first doing an app-code deploy, then doing a content-only deploy. But we can't do them both at the same time if there are accessibility issues on a Drupal page.

### Security Concerns

There are no new security concerns with a separated build process.

### Privacy Concerns

There are no new privacy concerns with a separated build process.

### Open Questions and Risks

#### Risks

> - Neither the current content validation nor the proposed content validation will **prevent** invalid content from making it to production. But this proposal would prevent unrelated teams from being blocked by invalid content.

Reviewer 6: This might be me being picky, but I'm not sure this is 100% true.  When you say "production" here, are you talking about the live va.gov site, or do you mean the CMS production server?  We have had incidents of bad content breaking the content build, which prevents new content from going to prod on va.gov

#### Questions

### Work Estimates

The following estimates vary greatly depending on who's doing the work.

1. Finish extracting the `content-build` repo out of the `vets-website` repo - ?
2. Write and test a scheduled job to run the content build - <1 day

### Alternatives

#### Highest quality, but lowest speed

If we could proactively validate CMS content before it's published, we could eliminate a source of downstream broken links and/or accessibility errors, which would improve the experience for end users and/or reduce the number of frontend deploys that require manual intervention. The downside of this approach would be figuring out how to add the content validation to the CMS, and/or adding friction to the content writers workflow.

#### Fastest speed, but lowest quality

> We could stop breaking the `vets-website` build due to invalid content. That would reduce the burden on the content writers and/or reduce the number of deploys requiring manual intervention. However, that would likely mean end users experience broken links and/or accessibility errors more often.

Reviewer 6: This line seems like it can be interpreted as contradicting an earlier statement:

> We could stop breaking the `vets-website` build due to invalid content.

The previous statement:
> Neither the current content validation nor the proposed content validation will **prevent** invalid content from making it to production.

When violations are found, the build breaks and violations must be reported manually.

The accessibility checks are skipped during a content-only deploy.

Additionally, we have the preview server which shows content editors any accessibility violations, but they must deliberately go to the server from a link in Drupal to see this.

#### Pain points

##### Content Writers

Content writers want to quickly draft, publish, and deploy content. However, `vets-website` is only deployed to production once per day. To allow content writers to move faster, there is a [partial deploy](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/getting-started/workflow/deploy/#partial-deploy--static-page-changes-only), which only includes static page changes (`vagov-content` and Drupal).

Unfortunately, the content-only deploys have no accessibility checking. That means that content writers don't have a proactive way to learn if their content includes accessibility errors. Adding accessibility checking to the CMS is on the roadmap, but it doesn't currently exist.

> ##### Users

Reviewer 5: ```suggestion
##### VA.gov Users
```

To enable a faster publishing workflow, all tests (except link checking) are intentionally skipped during content-only deploys. That creates an opportunity for end users to encounter accessibility errors. Additionally, inconsistencies between how the CMS handles redirects and how `vets-website` handles redirects, it's possible for users to encounter broken links in production.

##### FE Tools team

> Invalid content causes warning logs during staging deploys, and breaks the build during production deploys. These broken builds then need manual intervention, and require coordinating with another team.

Reviewer 5: ```suggestion
Invalid content causes warning logs during staging deploys, and breaks the build during production deploys. These broken builds then need manual intervention, and require coordinating with the CMS team.
```

##### Anybody that is trying to ship code for the VA.gov front end

If invalid content is breaking the production build, and therefore blocking the production deploy, then no team on any contract is able to ship their frontend code to VA.gov.

## High Level Design

> Once the `content-build` repo is successfully extracted from `vets-website`, the Metalsmith static content build pipeline will be deleted from `vets-website`. Webpack will handle the application build in `vets-website`, and Metalsmith will handle the static content build in the `content-build` repo. This will create two separate pipelines for built code to be deployed to S3, which will prevent failures in one from blocking the other.

Reviewer 5: ```suggestion
As part of the [build separation wor](https://github.com/department-of-veterans-affairs/va.gov-team/issues/2719)k, two separate pipelines will be created for for built code to be deployed to S3 to prevent failures in one from blocking between CMS and application deployments. 

The Metalsmith static content build will be extracted to the `content-build` repo and deleted from `vets-website`, and its deployment pipeline. Webpack will handle the application build for the `vets-website` repo, and Metalsmith will handle the static content build for the `content-build` repo. 
```

> The validation of static content will take place in a scheduled job that runs every workday. Since the broken link check happens during the Metalsmith build, and the accessibility check relies on the output of the content build, the scheduled job will run a script to produce a build that only has the steps necessary for content validation. That build will only be used for reporting broken links and accessibility errors, and will have no impact on deploys. Those reports will be delivered to the #cms-team channel via existing Slack integrations.

Reviewer 5: ```suggestion
The validation of static content will take place in a scheduled job that runs every workday instead of ???. Since the broken link check happens during the Metalsmith build, and the accessibility check relies on the output of the content build, the scheduled job will run a script to produce a build that only has the steps necessary for content validation. That build will only be used for reporting broken links and accessibility errors, and will have no impact on deploys. Those reports will be delivered to the #cms-team channel via existing Slack integrations.
```

Reviewer 5: This is what we're making a decision about on this document so we should emphasize that somehow.

## Specifics

### Detailed Design

Using the current content build script as a reference, we will write another build script that only includes the steps necessary for the content validation. We will reuse the existing broken link check from the current content build. And using the existing accessibility tests as a reference, we will configure the accessibility tests to not stop on failure so we can identify all the accessibility errors. The script will generate a report of all the broken links and all the accessibility errors. Those reports will be sent to the #cms-team channel, just like the current broken link notifications today.

Lastly, we will write a scheduled Jenkins workflow to run the content validation script every workday at midnight.

### Code Location

The content validation script and scheduled job config will live in the [new `content-build` repo](https://github.com/department-of-veterans-affairs/content-build).

### Testing Plan

1. Once the content validation script is written, we can run the script locally.
2. Once the script is behaving correctly in local environment, we can call the script in a scheduled job.

> 3. Once the script is being called in the scheduled job, we can monitor and optimize it.

Reviewer 5: We should add some more details to this. We should at least do positive and negative testing (locally and in the CI) and validate that errors are pushed to the right notification channels. Describe how you intend to do the different types of testing.

### Logging

Currently, invalid content is logged in the build log during staging deploys. And failed builds trigger generic [Slack notifications](https://dsva.slack.com/archives/C37M86Y8G/p1591099447374800).

The updated content validation approach will continue logging invalid content errors in the build log, and will also post an automated report of the invalid content in the #cms-team Slack channel.

### Debugging

##### Broken links

> 1. Confirm that the content build is passing

Reviewer 5: What can be used for this? Jenkins logs? Slack notifications?

2. Modify an HTML file, like [`header.html`](https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/includes/header.html), to have a broken link

> 3. Run the build again to confirm that it is failing

Reviewer 5: this sounds like a test plan. debugging should probably just detail how to find and interpret the errors in jenkins: 

>How users can debug interactions with your system. When designing a system it's important to think about what tools you can provide to make debugging problems easier. Sometimes it's unclear whether the problem is in your system at all, so a mechanism for isolating a particular interaction and examining it to see if your system behaved as expected is very valuable. Once a system is in use, this is a great place to put tips and recipes for debugging. If this section grows too large, the mechanisms can be summarized here and individual tips can be moved to another document.

##### Accessibility errors

1. Confirm that the content build is passing
2. Modify an HTML file, like [`header.html`](https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/includes/header.html), to have an accessibility error

> 3. Run the build again to confirm that it is failing

Reviewer 5: This also sounds like a test plan.

### Caveats

> - The current broken link checking in Drupal can handle redirects, but those redirects are not currently synchronized with `vets-website`. When the redirects are not synchronized between `vets-website` and Drupal, a page that is moved in Drupal may pass the Drupal broken link check but fail the `vets-website` broken link check.

Reviewer 5: I don't understand why this is a caveat. anchor tags from VA's CMS shouldn't point to redirects- redirects are for external referrers i.e. content we don't control.

Author: I added this based on [a comment from Steve W](https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658#discussion_r434864015). 

Do you think this paragraph needs to be moved (perhaps to the **Background** section), is incorrect, is confusing, and/or something else?

Reviewer 7: Re: whether or not this caveat needs to be here

Does this affect your design doc's plans in any way? It's true information, but I don't think it's relevant to the changes you're proposing in this doc.

Author: It doesn't affect the design doc's plans in any way. I've moved it from caveats to background in https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658/commits/4efc42728ba68c5c467fded098d3791f73d635fa.

No accessibility checking is performed on the content-only deploys. That means that content writers don't have a proactive way to learn if their content includes accessibility errors. Adding accessibility checking to the CMS is on the roadmap, but it doesn't currently exist.

##### VA.gov Users

To enable a faster publishing workflow, all tests (except link checking) are intentionally skipped during content-only deploys. That creates an opportunity for end users to encounter accessibility errors. Additionally, inconsistencies between how the CMS handles redirects and how `vets-website` handles redirects, it's possible for users to encounter broken links in production.

##### FE Tools team

Invalid content causes warning logs during staging deploys, and breaks the build during production deploys. These broken builds then need manual intervention, and require coordinating with the CMS team.

##### Anybody that is trying to ship code for the VA.gov front end

If invalid content is breaking the production build, and therefore blocking the production deploy, then no team on any contract is able to ship their frontend code to VA.gov.

## High Level Design

> As part of the [build separation wor](https://github.com/department-of-veterans-affairs/va.gov-team/issues/2719)k, two separate pipelines will be created for for built code to be deployed to S3 to prevent failures in one from blocking between CMS and application deployments.

Author: > wait, is this "two separate deploy pipelines" what you're proposing with this new design? or is this part of the context of other work already happening?

The two separate pipelines I'm referring to are part of the context of other work already happening. When I say "two separate deploy pipelines," I'm referring to the distinct content and application builds from [Chris' Separate Content Build design doc](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/engineering/design-docs/2020-04-09-separate-content-build.md#high-level-design).

The Metalsmith static content build will be extracted to the `content-build` repo and deleted from `vets-website`, and its deployment pipeline. Webpack will handle the application build for the `vets-website` repo, and Metalsmith will handle the static content build for the `content-build` repo.

> The validation of static content will take place in a scheduled job that runs every workday instead of ???. Since the broken link check happens during the Metalsmith build, and the accessibility check relies on the output of the content build, the scheduled job will run a script to produce a build that only has the steps necessary for content validation. That build will only be used for reporting broken links and accessibility errors, and will have no impact on deploys. Those reports will be delivered to the #cms-team channel via existing Slack integrations.

Author: Because the content-only deploy, AKA [partial deploy](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/getting-started/workflow/deploy#partial-deploy--static-page-changes-only), only runs the drupal broken link checks, and none of the `vets-website` broken link checks, we've already indirectly decided that fast content deploys are worth the risk of broken links landing in production. 

In the **Alternatives** section, I've included a [**Highest quality, but lowest speed** option](https://github.com/department-of-veterans-affairs/va.gov-team/blob/cfae598310780e75243d3b5f6cbad8f1031079aa/platform/engineering/design-docs/2020-05-26-content-validation.md#highest-quality-but-lowest-speed).

## Specifics

> ### Detailed Design

Reviewer 7: this makes sense as a starting point, but I think needs a _lot_ more detail.

each of the sentences you have below should probably have their own heading describing that decision and explanations about how to implement each, and maybe describing how they affect the pain points. but at the moment there isn't enough described to feel like this really communicates what's happening.

some initial questions that come to mind as examples of what detail this still needs:

> Using the current content build script as a reference, we will write another build script that only includes the steps necessary for the content validation. 

* Does that mean you'll remove content validation from the current content build script?
* What runs this new build script?
* What happens when this new build script detects invalid content?
* Will invalid content detection block a build or deployment?

> We will reuse the existing broken link check from the current content build. 

* does that mean just the checking logic? what about how reporting is done?

> And using the existing accessibility tests as a reference, we will configure the accessibility tests to not stop on failure so we can identify all the accessibility errors.

* Wait, are you using the accessibility tests as a reference for other work? Or changing the accessibility tests? This sounds like both, but to me those are mutually exclusive things
* What does "not stop on failure" mean --- just making sure we check for _all_ errors instead of stopping at the first error? Or does it mean not stopping a build/deploy?

To be clear, these are all separate questions that I feel like the design doc doesn't answer currently.

Reviewer 7: A few other questions I have reading this:

* Where will this script run? (Jenkins? CircleCI? somewhere else?)
* How often will it run? (you said "every workday" above, does that mean once/day? if it's just giving us notifications, should we run it more frequently?)
* What content will it run against? Fetched directly from Drupal? Running against what's live on the website?

Reviewer 7: Thanks for the extra detail! I feel much more able to understand your proposal now.

I didn't find an answer to this question of mine though:

> What content will it run against? Fetched directly from Drupal? Running against what's live on the website?

Author: > Thanks for the extra detail! I feel much more able to understand your proposal now.
> 
> I didn't find an answer to this question of mine though:
> 
> > What content will it run against? Fetched directly from Drupal? Running against what's live on the website?

@gunsch, sorry for missing that question. To answer your question, we will first try to pull the latest content from Drupal. If we are unable to pull the latest content from Drupal, we will use the cached Drupal content.

I've updated the design doc to include that answer in https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658/commits/6077200b080141dd0b7946cd6060fb54ff9e42ee.

##### Negative

1. Confirm that the content build has zero accessibility errors and zero broken links
2. Add a broken link, and an accessibility error
3. Run the job that calls the content validation script
4. Confirm that a report is sent to the #cms-team Slack channel
5. Confirm that the report includes the newly introduced accessibility error
6. Confirm that the report includes the newly introduced broken link
7. Fix the newly introduced accessibility error
8. Fix the newly introduced broken link
9. Run the job that calls the content validation script
10. Confirm that the report includes zero accessibility errors and zero broken links

### Logging

Currently, invalid content is logged in the build log during staging deploys. And failed builds trigger generic [Slack notifications](https://dsva.slack.com/archives/C37M86Y8G/p1591099447374800).

The updated content validation approach will continue logging invalid content errors in the build log, and will also post an automated report of the invalid content in the #cms-team Slack channel.

### Debugging

- Local build logs
- Jenkins build logs
- Slack notifications

### Caveats

> - The current broken link checking in Drupal can handle redirects, but those redirects are not currently synchronized with `vets-website`. When the redirects are not synchronized between `vets-website` and Drupal, a page that is moved in Drupal may pass the Drupal broken link check but fail the `vets-website` broken link check.

Reviewer 7: what does it mean to say "can handle redirects, but those redirects are not currently synchronized with `vets-website`"?

* does "handle redirects" mean "follow all 301/302s on the live site"? (if so, I don't understand the "synchronized with vets-website" comment)
* does it mean "understands redirects configured within Drupal"? (if so, could you add that?)

Author: It means that the redirects from drupal are not being imported into nginx during the content build. 

See the following quote from Steve W. for more info. 
> Where it still has the potential to fall down right now is that if a path to a page in drupal is changed, drupal automatically creates a redirect from the old path to the new path, so when the links are checked, they still work. The original plan was that the Front end would import redirects from drupal into Nginx when the content build takes place. However that is not happening, so the front end does not have the redirects, so a moved page, results in broken links from pre-existing content links. - https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658#discussion_r434864015

I'll update this section to be more clear.

      - [New scheduled job in Jenkins](#new-scheduled-job-in-jenkins)
    - [Code Location](#code-location)
    - [Testing Plan](#testing-plan)
      - [Local](#local)
        - [Broken links](#broken-links)
          - [Positive](#positive)
          - [Negative](#negative)
        - [Accessibility errors](#accessibility-errors)
          - [Positive](#positive-1)
          - [Negative](#negative-1)
      - [CI](#ci)
        - [Positive](#positive-2)
        - [Negative](#negative-2)
    - [Logging](#logging)
    - [Debugging](#debugging)
    - [Caveats](#caveats)
    - [Security Concerns](#security-concerns)
    - [Privacy Concerns](#privacy-concerns)
    - [Open Questions and Risks](#open-questions-and-risks)
      - [Risks](#risks)
      - [Questions](#questions)
    - [Work Estimates](#work-estimates)
    - [Alternatives](#alternatives)
      - [Highest quality, but lowest speed](#highest-quality-but-lowest-speed)
      - [Fastest speed, but lowest quality](#fastest-speed-but-lowest-quality)
    - [Future Work](#future-work)
    - [Revision History](#revision-history)

## Overview

### Objective

> The objective here is to define an updated approach to validate the CMS content. Specifically, this design doc focuses on identifying accessibility errors and/or broken links coming from the CMS content.

Author: I made the following change to the objective in https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658/commits/76febfe19d753aa2535ef231ad8e29d4d9359090. 

What do you think of this?
```
- The objective here is to define an updated approach to validate the CMS content. Specifically, this design doc focuses on identifying accessibility errors and/or broken links coming from the CMS content.
+ The objective here is to ensure that CMS content is still being validated once the `content-build` has been successfully extracted from the `vets-website` repo.
```

> The objective here is to define an updated approach to validate the CMS content. Specifically, this design doc focuses on identifying accessibility errors and/or broken links coming from the CMS content.

Reviewer 7: Awesome! Thanks, even just that small change helps me put the rest of the doc in a clearer context.

##### FE Tools team

Invalid content causes warning logs during staging deploys, and breaks the build during production deploys. These broken builds then need manual intervention, and require coordinating with the CMS team.

##### Anybody that is trying to ship code for the VA.gov front end

If invalid content is breaking the production build, and therefore blocking the production deploy, then no team on any contract is able to ship their frontend code to VA.gov.

## High Level Design

Once the [build separation work](https://github.com/department-of-veterans-affairs/va.gov-team/issues/2719) is complete, there will be separate content and application builds. These distinct builds will prevent failures in one from blocking deployments to the other.

The Metalsmith static content build will be extracted to the `content-build` repo and deleted from the `vets-website` repo and deployment pipeline. Webpack will handle the application build for the `vets-website` repo, and Metalsmith will handle the static content build for the `content-build` repo.

> The validation of static content will take place in a scheduled job that runs every workday, instead of happening as part of the full deploy. Since the broken link check happens during the Metalsmith build, and the accessibility check relies on the output of the content build, the scheduled job will run a script to produce a build that only has the steps necessary for content validation. That build will only be used for reporting broken links and accessibility errors, and will have no impact on deploys. Those reports will be delivered to the #cms-team channel via existing Slack integrations.

Reviewer 7: > That build will only be used for reporting broken links and accessibility errors, and will have no impact on deploys.

is this still true?

Author: That is no longer true. Now that we have determined that broken links do block content only deploys, I have removed the mention of broken links. That build will now only be used for reporting accessibility errors, and that build will have no impact on deploys.

For anybody trying to ship code for the VA.gov front end, this approach will eliminate a source of failed application builds, which should allow them to ship code faster.

For VA.gov users, this approach will continue allowing users to quickly see updated content. The downside of this approach for VA.gov users is that it won't eliminate existing opportunities for users to see accessibility errors. Fortunately, the improved error reporting should allow content writers to easily identify, remedy, and deploy valid content, which should minimize the user's chance of encountering invalid content.

In summary, this approach will have the following impact for the following groups:

- **Content Writers**: Improve error reporting while maintaining current convenience
- **Frontend Tools team**: Reduce build noise, failure rate, and number of builds requiring manual intervention
- **VA.gov users**: Maintain existing time to see updated content and risk of seeing accessibility errors

> - **Anybody trying to ship code for the VA.gov front end**: Reduce failed deploys which should allow them to ship faster

Reviewer 7: given our conversations about broken link checking still blocking deploys after all... what builds will we no longer be failing?

Author: With this approach, the application build will no longer fail due to accessibility errors. 

Phrased another way:
**Current**: a11y content errors break the application build, but not the content build
**Proposed**: a11y content errors will not break the application build, and continue not breaking the content build

#### Update accessibility tests config

The [current accessibility tests](https://github.com/department-of-veterans-affairs/vets-website/blob/f93e0c11d9e037bd2460ede5e2fb4b67fbaaf118/Jenkinsfile#L82-L84) use Nightwatch. The default value for [Nightwatch's `end_session_on_fail` test session setting](https://nightwatchjs.org/gettingstarted/configuration/#test-session-settings) is `true`. Since we want a report of all the failures, we would need to update the `end_session_on_fail` setting in our [Nightwatch config](https://github.com/department-of-veterans-affairs/content-build/blob/master/config/nightwatch.docker-compose.js) from `true` to `false`.

#### Improved invalid content reporting

Currently, accessibility errors need to be manually communicated from the Frontend Tools team to the CMS team. And broken links need to be extracted from the build log. We would improve that reporting by generating a file using Winston, and send that file to the #cms-team channel using existing Slack integrations.

#### New scheduled job in Jenkins

> Lastly, we will write a scheduled Jenkins workflow to run the content validation script hourly during workdays. That way, the content team learns about invalid content at a convenient cadence.

Reviewer 7: Can we run scheduled jobs in CircleCI? (i.e. what do we do post-Jenkins, since Ops team is tearing that down before too much longer?)

Author: CircleCI does support scheduled workflows. 

https://circleci.com/docs/2.0/workflows/#scheduling-a-workflow

I actually attempted a prototype of a CircleCI scheduled job in a [branch](https://github.com/department-of-veterans-affairs/vets-website/compare/add-nightly-build-to-circleci) a couple months ago.

Reviewer 7: great! is there a reason we wouldn't just implement this in CircleCI to start?

Author: 🤔 

On one hand, I can't cite a concrete reason off the top of my head. 

On the other hand, I could see an argument that switching CI tools is slightly outside the scope/spirit of this design doc, and that task would be a good candidate for the **Future Work** section.

Reviewer 7: my 2c: it seems like a strong likelihood that we're going to switch fully to CircleCI. so options are:
- implement this in Jenkins now, have to move it to CircleCI later
- implement in CircleCI, don't have to move later
if that's the case, should we just implement in Circle right away and save the trouble?

Author: That makes sense. I've updated the Detailed Design to specify CircleCI instead of Jenkins in https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658/commits/bdc89d7347e9ca7df9ac23524d52eaf848fafe3e.

### Code Location

The broken link checks, axe checks, and scheduled job config will live in the [new `content-build` repo](https://github.com/department-of-veterans-affairs/content-build).

> ### Testing Plan

Reviewer 7: Btw, it's OK for testing plan to be much more high-level than this! This is extremely detailed, and I think just saying we'd:
* Validate that broken link checking still works
* Validate that accessibility validation still works
* Validate that accessibility validation on CI sends alerts to #cms-team 

is probably sufficient. (No need to change on this doc, just a note for future docs!)

6. Confirm that the report includes the newly introduced broken link
7. Fix the newly introduced accessibility error
8. Fix the newly introduced broken link
9. Run the job that calls the content validation script
10. Confirm that the report includes zero accessibility errors and zero broken links

### Logging

Currently, invalid content is logged in the build log during staging deploys. And failed builds trigger generic [Slack notifications](https://dsva.slack.com/archives/C37M86Y8G/p1591099447374800).

The updated content validation approach will continue logging invalid content errors in the build log, and will also post an automated report of the invalid content in the #cms-team Slack channel.

### Debugging

- Local build logs
- Jenkins build logs
- Slack notifications

### Caveats

- The current broken link checking in Drupal can handle redirects, but those redirects are not currently synchronized with `vets-website`. When the redirects are not synchronized between `vets-website` and Drupal, a page that is moved in Drupal may pass the Drupal broken link check but fail the `vets-website` broken link check.

> - Many of the broken links come from URLs that change, according to decisions by the Content & IA team, usually in conjunction with Public Websites. In that workflow, an engineer makes a code change for the redirect, and has to time the release of the redirect code to match the content release of the URL change in the content layer. It's a fragile and very time-consuming process, and will likely not be scalable as we add hundreds of editors from outside the tight Github/Slack communication flows (e.g. Public Affairs Officers at 2000+ VA facilities).

Reviewer 7: (nit) Also doesn't really feel like a caveat, maybe background info?

Author: Moved from caveat to background in https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658/commits/152e818308af23b7148b4acbec1de630e52fe6fd

4. Blocks the deployment on production or logs the error output on lower environments

Code Path for the `check-broken-links` Metalsmith plugin

1. `package.json`’s `build-content` script gets called
2. `script/build-content.js` calls the build script for Metalsmith
3. `src/site/stages/build/index.js` calls the `check-broken-links` plugin
4. `src/site/stages/build/plugins/check-broken-links/index.js` calls the `getBrokenLinks` helper
5. `src/site/stages/build/plugins/check-broken-links/helpers/getBrokenLinks.js` calls the `isBrokenLink` function
6. `src/site/stages/build/plugins/check-broken-links/helpers/isBrokenLink.js` checks if an internal `href`/`src` value is included a specific array of paths
7. The `check-broken-links` plugin calls the `getBrokenLinks` helper to build up an array of broken links
8. The array of broken links from the `getBrokenLinks` helper is used by the `getErrorOutput` helper to build a CSV of broken links
9. The `check-broken-links` plugin logs the broken links in the build console

![screenshot of code path for check-broken-links Metalsmith plugin](https://user-images.githubusercontent.com/6130520/87714554-8849ef00-c771-11ea-8f20-ed52af7fcd3a.png)

> If there are broken links, the `glean-broken-links` script the build log, and create a file of broken links in a Comma-Separated Value (CSV) format. If broken links are found on the `master` branch, then an exception is thrown during the Jenkins job to block the deploy.

Reviewer 8: ```suggestion
If there are broken links, the `glean-broken-links` script parses the build log and creates a file that lists the broken links in a Comma-Separated Value (CSV) format. If broken links are found on the `master` branch, an exception is thrown during the Jenkins job to block the deploy. 
```

##### Anybody that is trying to ship code for the VA.gov front end

If invalid content is breaking the production build, and therefore blocking the production deploy, then no team on any contract is able to ship their frontend code to VA.gov.

## High Level Design

Once the [build separation work](https://github.com/department-of-veterans-affairs/va.gov-team/issues/2719) is complete, there will be separate content and application builds. These distinct builds will prevent failures in one from blocking deployments to the other.

The Metalsmith static content build will be extracted to the `content-build` repo and deleted from the `vets-website` repo and deployment pipeline. Webpack will handle the application build for the `vets-website` repo, and Metalsmith will handle the static content build for the `content-build` repo.

The broken link validation will continue using the existing the `check-broken-link` Metalsmith plugin and `glean-broken-links` script. Since broken links currently block content-only deploys, then we will continue blocking content-only deploys when broken links are detected.

> The validation of accessibility errors will take place in a scheduled job that runs hourly during workdays, instead of happening as part of the full deploy. Since the accessibility check relies on the output of the content build, the scheduled job will run a script to produce a build that only has the steps necessary for accessibility checking. That build will only be used for reporting accessibility errors, and will have no impact on deploys. Those reports will be delivered to the #cms-team channel via existing Slack integrations.

Reviewer 9: I saw above that there's a plan to get a11y testing into the CMS, maybe it's worth accelerating that so the site doesn't risk falling out of section 508/WCAG AA compliance? I think code reviews should mitigate it from the apps end, but content I could see tripping up quite a bit - maybe if it's just trying to recommend a Drupal plugin or something (cc @meganhkelley)

Reviewer 10: Accessibility checks for content on VA.gov are split over a variety of places — part of the CMS, part of automated tests on vets-website, and as manual reviews done before and after products and features launch. Features within the CMS itself would fall under purview of the CMS team — but we currently maintain the CMS build pipeline, which has some accessibility checks, which is why it came up in this design doc. 

In other words, let's chat more!

Reviewer 2: The issue is most of the content from the CMS is in fields that lack the bigger html wrappers, so it can not be accurately tested since it is not rendered the same way it is by the  metalsmith build.  

There are some WYSIWIG instances in the cms but those are fairly restricted as far as what html tags people can use and what they can add.  It is not a guarantee that what comes out of them is accessible, but it is often the case that what may be needed for accessiblity may not be attainable in the wysiwig.    There is some scanning we could do and should do, but it will not be the panacea we want it to be.

Reviewer 9: we decided that since the a11y tests take 30mins to run, hourly didn't makes sense, so we're backing this off to daily

#### Improved invalid content reporting

Currently, accessibility errors need to be manually communicated from the Frontend Tools team to the CMS team. And broken links need to be extracted from the build log. We would improve that reporting by generating a file using Winston, and send that file to the #cms-team channel using existing Slack integrations.

#### New scheduled job in CircleCI

Lastly, we will write a scheduled workflow to run the content validation script hourly during workdays. That way, the content team learns about invalid content at a convenient cadence.

Since we plan on migrating from Jenkins to CircleCI in the future, we will write this new job in CircleCI.

### Code Location

The broken link checks, axe checks, and scheduled job config will live in the [new `content-build` repo](https://github.com/department-of-veterans-affairs/content-build).

### Testing Plan

#### Local

##### Broken links

###### Positive

1. Confirm that the content build has zero broken links
2. Run the content validation script

> 3. Confirm that the script reports zero broken links

Reviewer 11: Aside from broken links, is there any plan to validate that the content is rendered correctly?

On the facilities team, we have started using [Cypress e2e tests](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/layouts/tests/vamc/smoke.mobile.cypress.spec.js) to validate our work.
Will this approach be supported going forward?

###### Negative

1. Confirm that the content build has zero broken links
2. Modify an HTML file, like [`header.html`](https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/includes/header.html), to have a broken link
3. Run the content validation script
4. Confirm that the script reports the newly introduced broken link

##### Accessibility errors

###### Positive

1. Confirm that the content build has zero accessibility errors
2. Run the content validation script
3. Confirm that the script reports zero accessibility errors

###### Negative

1. Confirm that the content build has zero accessibility errors
2. Modify an HTML file, like [`header.html`](https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/includes/header.html), to have an accessibility error
3. Confirm that the script reports the newly introduced accessibility error

#### CI

##### Positive

1. Confirm that the content build has zero accessibility errors and zero broken links

> 2. Run the job that calls the content validation script

Reviewer 11: Aside from broken links, is there any plan to validate that the content is rendered correctly?

On the facilities team, we have started using [Cypress e2e tests](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/layouts/tests/vamc/smoke.mobile.cypress.spec.js) to validate our work.
Will this approach be supported going forward?

Reviewer 9: @mpelzsherman nothing currently, but I'm happy to run/take a look at anything anyone has

9. Run the job that calls the content validation script
10. Confirm that the report includes zero accessibility errors and zero broken links

### Logging

Currently, invalid content is logged in the build log during staging deploys. And failed builds trigger generic [Slack notifications](https://dsva.slack.com/archives/C37M86Y8G/p1591099447374800).

The updated content validation approach will continue logging invalid content errors in the build log, and will also post an automated report of the invalid content in the #cms-team Slack channel.

### Debugging

- Local build logs
- Jenkins build logs
- CircleCI build logs
- Slack notifications

### Caveats

### Security Concerns

There are no new security concerns with a separated content validation process.

### Privacy Concerns

There are no new privacy concerns with a separated content validation process.

### Open Questions and Risks

#### Risks

- Neither the current content validation nor the proposed content validation will **prevent** invalid content from making it to the live VA.gov site. But this proposal would prevent unrelated teams from being blocked by invalid content.

#### Questions

> - What is the optimal frequency to run the scheduled job? Daily seems too slow. Every 10 minutes might be too fast for our builds. Hourly seems feasible. Is hourly acceptable? Should we be c concern about noise from a higher frequency?

Reviewer 9: I think hourly is OK, but errors should pile up. Like, if one error reports and isn't fixed it should be reported again after an hour and keep reporting until it is fixed

Reviewer 2: I'd suggest that this should run with every content release.  Otherwise the reporting could be completely out of sync with what is actually out there and have editors thrashing to chase down reported issues they already resolved.

Reviewer 9: Yeah, I agree there, but I would also scan hourly in case there's a large gap in between content releases. So something doesn't sit there for too long and risk getting missed.

Author: The hourly idea came from the previous platform lead:

>how long do we anticipate this job taking? could we run it more frequently? (hourly? every 10 minutes?) - https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658#discussion_r458442981

#### Fastest speed, but lowest quality

Since the partial deploy already provides an avenue for invalid content to land on the live VA.gov site, we could remove all the content validation from the full `vets-website` build. That would reduce the burden on the content writers and/or reduce the number of deploys requiring manual intervention. However, that would likely mean end users experience broken links and/or accessibility errors more often.

That would effectively be downgrading invalid content from an error to a warning.

### Future Work

- Delete the `accessibility` step of the `integration` stage from the `vets-website` Jenkins pipeline because that step will no longer be necessary in `vets-website`. It will no longer be necessary because the application accessibility checks are included in the tests run by the `e2e` step.

### Revision History

| Date         | Revisions Made | Author        |
| ------------ | -------------- | ------------- |

> | May 26, 2020 | Initial draft  | Bill Fienberg |

Reviewer 11: There have been many commits to this document since May 26, so this section is misleading. Consider deleting this since the revision history can be obtained from git?

Author: I see what you mean. Initially, I included more entries in the revision history, but I removed them per the following PR comment:

>I think you should only need to add revision history after the document has been approved and merged. - https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658#discussion_r433466403

For anybody trying to ship code for the VA.gov front end, this approach will eliminate a source of failed application builds, which should allow them to ship code faster.

For VA.gov users, this approach will continue allowing users to quickly see updated content. The downside of this approach for VA.gov users is that it won't eliminate existing opportunities for users to see invalid content. Fortunately, the improved error reporting should allow content writers to easily identify, remedy, and deploy valid content, which should minimize the user's chance of encountering invalid content.

In summary, this approach will have the following impact for the following groups:

- **Content Writers**: Improve error reporting while maintaining current convenience
- **Frontend Tools team**: Reduce build noise, failure rate, and number of builds requiring manual intervention

> - **VA.gov users**: Maintain existing time to see updated content and risk of seeing invalid content

Reviewer 7: hmm, is this true? my understanding was that broken link checking currently gates deployment, so if we move it to a daily job that _doesn't_ gate deployment, doesn't that increase risk of seeing invalid content?

Author: Your understanding is correct. I apologize for the confusion. The current system does block deploys when broken links are detected in `master`. I've updated the design doc to clarify this in https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658/commits/c3389e0b45f31d14079cc1a8d26a898d0a280de8.

We will reuse the existing `check-broken-link` Metalsmith plugin in the new build script.

#### Update accessibility tests config

The [current accessibility tests](https://github.com/department-of-veterans-affairs/vets-website/blob/f93e0c11d9e037bd2460ede5e2fb4b67fbaaf118/Jenkinsfile#L82-L84) use Nightwatch. The default value for [Nightwatch's `end_session_on_fail` test session setting](https://nightwatchjs.org/gettingstarted/configuration/#test-session-settings) is true. Since we want a report of all the failures, we would need to update the `end_session_on_fail` setting in our [Nightwatch config](https://github.com/department-of-veterans-affairs/content-build/blob/master/config/nightwatch.docker-compose.js) from `true` to `false`.

#### Improved invalid content reporting

Currently, accessibility errors need to be manually communicated from the Frontend Tools team to the CMS team. And broken links need to be extracted from the build log. We would improve that reporting by generating a file using Winston, and send that file to the #cms-team channel using existing Slack integrations.

#### New scheduled job in Jenkins

> Lastly, we will write a scheduled Jenkins workflow to run the content validation script every workday at midnight Eastern. That way, the CMS team should have an accurate report of invalid content when they start their workday.

Reviewer 7: how long do we anticipate this job taking? could we run it more frequently? (hourly? every 10 minutes?)

Author: It looks like the existing `e2e` and `accessibility` steps, which run in parallel, take about 16 minutes to run. There's no reason I can currently think of why we couldn't run the job more frequently. I think hourly makes sense. Do you have any concerns about noise from an increased frequency?

Reviewer 7: The tradeoff seems to be "increased noise" vs "increased time to catch a Veteran-facing issue", right?

If that's the case, I'm inclined to lean more toward "increased noise". We can always come up with creative ways to deal with noise (i.e. some extra logic to avoid alerting multiple times about the same issue), not as much for user-visible accessibility issues on the site waiting to be caught by validation.
